File: Solutions/Codeforces Solutions/1000_E.cpp
#include <bits/stdc++.h>
#include <bitset>

#define REP(i,n) for(int i=0;i<(int)n;i++)
#define REP1(i,j,n) for(int i=j;i<(int)n;i++)
#define all(x) x.begin(),x.end()
#define double long double
#define BUG cerr<<"BUG\n";

typedef long long ll;

using namespace std;

class bridgeTree{
public:
    void dfs();
    vector<vector<int> > G,T;
    vector<int>dfs_low,dfs_num,in_scc;
    stack<int>st;
    int n,counter,scc;
    bridgeTree(int nn,vector<int> *GG){
        n=nn;
        G.assign(n,vector<int>());
        REP(i,n)G[i]=GG[i];
        dfs_num.assign(n,0);
        dfs_low.assign(n,0);
        in_scc.assign(n,-1);
        scc=0;
        counter=0;
        for(int i=0;i<n;i++)
            if(dfs_num[i]==0)
                dfs(i);
        T.assign(scc,vector<int>());
        for(int i=0;i<n;i++)
            for(int v:G[i])
                if(in_scc[i]!=in_scc[v]){
                    T[in_scc[i]].push_back(in_scc[v]);
                    T[in_scc[v]].push_back(in_scc[i]);
                }
        for(int i=0;i<scc;i++){
            sort(all(T[i]));
            T[i].erase(unique(all(T[i])),T[i].end());
        }
    }
    void dfs(int u,int p=-1){
        dfs_num[u]=dfs_low[u]=++counter;
        st.push(u);
        for(int v:G[u])
            if(v!=p){
                if(dfs_num[v]==0)
                    dfs(v,u);
                if(in_scc[v]==-1)
                    dfs_low[u]=min(dfs_low[u],dfs_low[v]);
            }
        if(dfs_num[u]==dfs_low[u]){
            while(1){
                int v=st.top();st.pop();
                in_scc[v]=scc;
                if(v==u)break;
            }
            scc++;
        }
    }
};

int n;
vector<vector<int> >T;
const int N=3e5+10;
vector<int>G[N];

pair<int,int>dfs(int u,int p=-1){
    auto ret=make_pair(u,0);
    for(int v:T[u])
        if(v!=p){
            auto go=dfs(v,u);
            if(go.second+1>ret.second)
                ret=make_pair(go.first,go.second+1);
        }
    return ret;
};

int main(){
    ios_base::sync_with_stdio(0);
    cin.tie(0);

    int m;
    cin>>n>>m;
    while(m--){
        int u,v;
        cin>>u>>v;
        G[--u].push_back(--v);
        G[v].push_back(u);
    }
    bridgeTree bt(n,G);
    T=bt.T;
    cout<<dfs(dfs(0).first).second<<"\n";
}

File: Solutions/Codeforces Solutions/1000_E.java
import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.Stack;
import java.util.StringTokenizer;

/**
 * Created by omar on 5/19/18.
 */
public class A {

    static class Pair{
        int u,d;
        Pair(int u,int d){
            this.u=u;
            this.d=d;
        }
    }
    static ArrayList<Integer>[] adjList;
    static ArrayList<Integer>[] adjList3;
    static int V, counter, SCC, dfs_num[], dfs_low[];
    static int[] inSCC;
    static Stack<Integer> stack;


    static void tarjanSCC()
    {
        dfs_num=new int[V];
        dfs_low=new int[V];
        inSCC = new int[V];
        stack= new Stack<>();
        for(int i = 0; i < V; ++i)
            if(dfs_num[i] == 0)
                tarjanSCC(i);
        for(int i=0;i<V;i++)
            for(int v:adjList[i]){
                if(inSCC[i]==inSCC[v])continue;
                adjList3[inSCC[i]-1].add(inSCC[v]-1);
                adjList3[inSCC[v]-1].add(inSCC[i]-1);
            }
    }

    static void tarjanSCC(int u)
    {
        dfs_num[u] = dfs_low[u] = ++counter;
        stack.push(u);

        for(int v : adjList[u])
        {
            if(dfs_num[v] == 0)
                tarjanSCC(v);
            if(inSCC[v]==0)
                dfs_low[u] = Math.min(dfs_low[u], dfs_low[v]);
        }
        if(dfs_num[u] == dfs_low[u])
        {
            //SCC found
            while(true)
            {
                int v = stack.pop();
                inSCC[v] = SCC+1;
                if(v == u)
                    break;
            }
            SCC++;
        }
    }

    static ArrayList<Integer>tree[];
    static boolean vis[];
    static void dfs(int u,int p){
        vis[u]=true;
        for(int v:tree[u]){
            if(v==p)continue;
            adjList[u].add(v);
            if(!vis[v])
                dfs(v,u);
        }
    }
    static Pair dfs2(int u,int p){
        Pair ret=new Pair(u,0);
        for(int v:adjList3[u])
            if(v!=p){
                Pair go=dfs2(v,u);
                if(go.d+1>ret.d)
                    ret=new Pair(go.u,go.d+1);
            }
        return ret;
    };
    public static void main(String[] args) throws Exception {
        Scanner sc = new Scanner(System.in);
        PrintWriter out = new PrintWriter(System.out);

        V=sc.nextInt();
        tree = new ArrayList[V];
        adjList = new ArrayList[V];
        adjList3 = new ArrayList[V];
        vis = new boolean[V];
        for(int i=0;i<V;i++)tree[i]=new ArrayList<>();
        for(int i=0;i<V;i++)adjList[i]=new ArrayList<>();
        for(int i=0;i<V;i++)adjList3[i]=new ArrayList<>();
        int m=sc.nextInt();
        while(m-->0){
            int u=sc.nextInt()-1,v=sc.nextInt()-1;
            tree[u].add(v);
            tree[v].add(u);
        }
        dfs(0,-1);

        tarjanSCC();
        int far1=dfs2(0,-1).u;
        System.out.println(dfs2(far1,-1).d);
        out.flush();
    }

    static class Scanner {
        StringTokenizer st;
        BufferedReader bf;

        Scanner(InputStream in) {
            bf = new BufferedReader(new InputStreamReader(in));
        }

        String next() throws Exception {
            while (st == null || !st.hasMoreTokens()) {
                st = new StringTokenizer(bf.readLine());
            }
            return st.nextToken();
        }

        int nextInt() throws Exception {
            return Integer.parseInt(next());
        }

        long nextLong() throws Exception {
            return Long.parseLong(next());
        }
    }
}
